src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java

```diff
@@ -0,0 +1,159 @@
+package fi.aalto.cs.apluscourses.intellij.utils;
+
+import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.intellij.model.IntelliJModuleMetadata;
+import fi.aalto.cs.apluscourses.model.Module;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Paths;
+import java.time.ZonedDateTime;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.io.FileUtils;
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+public class CourseFileManager {
+
+  private File courseFile;
+  private URL courseUrl;
+  private Map<String, IntelliJModuleMetadata> modulesMetadata;
+
+  private CourseFileManager() {
+
+  }
+
+  private static final CourseFileManager instance = new CourseFileManager();
+
+  private static final String COURSE_FILE_NAME = "a-plus-project.json";
+
+  public static CourseFileManager getInstance() {
+    return instance;
+  }
+
+  /**
+   * Attempts to create a course file and load it for the given project with the given URL. If a
+   * course file for already exists (even with a different URL), this method loads the existing
+   * course file.
+   * @param project      The project for which the course file is created.
+   * @param courseUrl    The URL that gets added to the course file (if it doesn't exist yet).
+   * @throws IOException If an IO error occurs.
+   */
+  public synchronized void createAndLoad(@NotNull Project project, @NotNull URL courseUrl)
```

Comment:
it seems that now most of the IO operations are `syncronized`, do you think there could be alternative options for this?

Reply:
That's a good point. No alternatives come to mind at the moment, but we should keep it mind for the future :+1:  Luckily this is rarely used method.